tests/cpp/test_moe.cpp

```diff
@@ -0,0 +1,184 @@
+}
+
+TEST_F(SgLangMoETest, ComputeExpertOffsets) {
+  auto fusion_ptr = std::make_unique<Fusion>();
+  Fusion& fusion = *fusion_ptr.get();
+  FusionGuard fg(&fusion);
+
+  // tokens_per_expert
+  auto tv0 = makeContigConcreteTensor({num_experts}, DataType::Int);
+  fusion.addInput(tv0);
+
+  // Inclusive scan
+  auto tv1 = cumsum(tv0, 0);
+  // Exclusive scan + total count
+  auto tv2 =
+      pad(tv1, {fusion.oneVal(DataType::Int), fusion.zeroVal(DataType::Int)});
+
+  fusion.addOutput(tv2);
+
+  auto options = at::TensorOptions().dtype(at::kLong).device(at::kCUDA, 0);
+  auto t0 = at::randint(0, num_tokens * k, {num_experts}, options);
+
+  FusionExecutorCache executor_cache(std::move(fusion_ptr));
+  auto outputs = executor_cache.runFusionWithInputs({t0});
+
+  testValidate(&fusion, outputs, {t0}, __LINE__, __FILE__);
+}
+
+TEST_F(SgLangMoETest, ComputeExpertBlockScaleOffsets) {
+  auto fusion_ptr = std::make_unique<Fusion>();
+  Fusion& fusion = *fusion_ptr.get();
+  FusionGuard fg(&fusion);
+
+  // tokens_per_expert
+  auto tv0 = makeContigConcreteTensor({num_experts}, DataType::Int);
+  fusion.addInput(tv0);
+
+  // The first part is the same as ComputeExpertOffsets
+  auto tv1 = cumsum(tv0, 0);
+  auto tv2 =
+      pad(tv1, {fusion.oneVal(DataType::Int), fusion.zeroVal(DataType::Int)});
+
+  // Does the same cumsum with rounded token counts
+  auto tv3 =
+      ceilDiv(tv0, IrBuilder::create<Val>(rounding_factor, DataType::Int));
+  auto tv4 = mul(tv3, IrBuilder::create<Val>(rounding_factor, DataType::Int));
+  auto tv5 = cumsum(tv4, 0);
+  auto tv6 =
+      pad(tv5, {fusion.oneVal(DataType::Int), fusion.zeroVal(DataType::Int)});
+
+  fusion.addOutput(tv2);
+  fusion.addOutput(tv6);
+
+  auto options = at::TensorOptions().dtype(at::kLong).device(at::kCUDA, 0);
+  auto t0 = at::randint(0, num_tokens * k, {num_experts}, options);
+
+  FusionExecutorCache executor_cache(std::move(fusion_ptr));
+  auto outputs = executor_cache.runFusionWithInputs({t0});
+
+  // Manually computing the results as ceilDiv doesn't seem to work
+  // with ExprEvaluator. This is the error message:
+  //
+  // C++ exception with description "std::get: wrong index for variant" thrown
+  // in the test body.
+  auto t2 = at::pad(at::cumsum(t0, 0), {1, 0});
+  auto t0_rounded =
+      at::floor((t0 + rounding_factor - 1) / rounding_factor) * rounding_factor;
+  auto t6 = at::pad(at::cumsum(t0_rounded, 0), {1, 0});
+
+  testValidate(&fusion, outputs, {t0}, {t2, t6}, __LINE__, __FILE__);
+}
+
+TEST_F(SgLangMoETest, ComputeArgSort) {
+  auto fusion_ptr = std::make_unique<Fusion>();
+  Fusion& fusion = *fusion_ptr.get();
+  FusionGuard fg(&fusion);
+
+  // topk_ids
+  auto tv0 = makeContigConcreteTensor({num_tokens, k}, DataType::Int);
+  fusion.addInput(tv0);
+
+  auto tv1 = flatten(tv0);
+  // argsort indices
+  auto tv2 = argsort(tv1, 0, /*descending=*/true, /*stable=*/true);
+  // input_permutation
+  auto tv3 = div(tv2, IrBuilder::create<Val>(k, DataType::Int));
```

Comment:
naive question: why are we dividing by `k` here?

Reply:
That is to find the offset to the token dimension. Corresponds to https://github.com/sgl-project/sglang/blob/a7efbb2757721092732ffa3ab53b259e1d0ea83a/sgl-kernel/csrc/moe/prepare_moe_input.cu#L86